File: 백준/Gold/17837. 새로운 게임 2/새로운 게임 2.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Deque;
import java.util.List;
import java.util.StringTokenizer;

public class Main {

    static final int WHITE = 0;
    static final int RED = 1;
    static final int BLUE = 2;

    static int[] dx = {0, 0, -1, 1};
    static int[] dy = {1, -1, 0, 0};

    static int n;
    static int[][] board;
    static Horse[] horseArr;
    static Deque<Integer>[][] horseList;

    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        n = Integer.parseInt(st.nextToken());
        int k = Integer.parseInt(st.nextToken());

        board = new int[n][n];
        horseList = new ArrayDeque[n][n];

        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < n; j++) {
                board[i][j] = Integer.parseInt(st.nextToken());
                horseList[i][j] = new ArrayDeque<>();
            }
        }

        horseArr = new Horse[k];

        for (int i = 0; i < k; i++) {
            st = new StringTokenizer(br.readLine());

            int x = Integer.parseInt(st.nextToken()) - 1;
            int y = Integer.parseInt(st.nextToken()) - 1;
            int d = Integer.parseInt(st.nextToken()) - 1;

            horseArr[i] = new Horse(x, y, d);
            horseList[x][y].offer(i);
        }

        int turns = 0;

        while (turns <= 1000) {

            turns++;

            for (int i = 0; i < k; i++) {
                Horse h = horseArr[i];

                int x = h.x;
                int y = h.y;
                int dir = h.dir;

                Deque<Integer> stk = horseList[x][y];

                int nx = x + dx[dir];
                int ny = y + dy[dir];

                List<Integer> target = new ArrayList<>();
                while (!stk.isEmpty() && stk.peek() != i) {
                    target.add(stk.pop());
                }
                target.add(stk.pop());
                if (isBlocked(nx, ny)) {

                    dir = getOppositeDir(dir);
                    nx = x + dx[dir];
                    ny = y + dy[dir];

                    if (isBlocked(nx, ny)) {
                        if (horsesMove(target, x, y, false)) {
                            System.out.println(turns);
                            return;
                        }
                        horseArr[i].setDir(dir);
                    } else {
                        if (horsesMove(target, nx, ny, board[nx][ny] == RED)) {
                            System.out.println(turns);
                            return;
                        }
                        horseArr[i].setDir(dir);
                    }

                } else {
                    if (horsesMove(target, nx, ny, board[nx][ny] == RED)) {
                        System.out.println(turns);
                        return;
                    }
                }
            }
        }

        System.out.println(-1);
    }

    private static boolean horsesMove(List<Integer> target, int nx, int ny, boolean isRed) {

        if (isRed) Collections.reverse(target);

        for (int i = target.size() - 1; i >= 0; i--) {
            int targetIdx = target.get(i);
            horseArr[targetIdx].move(nx, ny);
            horseList[nx][ny].push(targetIdx);
        }

        return horseList[nx][ny].size() >= 4;
    }

    private static boolean isBlocked(int x, int y) {
        return x < 0 || y < 0 || x >= n || y >= n || board[x][y] == BLUE;
    }

    private static int getOppositeDir(int dir) {
        return (dir % 2 == 0) ? dir + 1 : dir - 1;
    }

    static class Horse {

        int x, y, dir;

        public Horse(int x, int y, int dir) {
            this.x = x;
            this.y = y;
            this.dir = dir;
        }

        public void move(int x, int y) {
            this.x = x;
            this.y = y;
        }

        public void setDir(int dir) {
            this.dir = dir;
        }
    }
}

File: 백준/Gold/17837. 새로운 게임 2/새로운 게임 2.py
from collections import deque

RED = 1
BLUE = 2

dx = (0, 0, -1, 1)
dy = (1, -1, 0, 0)


class Unit:
    def __init__(self, x, y, dir):
        self.x = x
        self.y = y
        self.dir = dir

    def move(self, x, y):
        self.x = x
        self.y = y

    def set_dir(self, dir):
        self.dir = dir


def is_blocked(x, y):
    return x < 0 or y < 0 or x >= n or y >= n or board[x][y] == BLUE


def horses_move(target, nx, ny, is_red):

    while target:
        unit_index = target.popleft() if is_red else target.pop()
        unit_arr[unit_index].move(nx, ny)
        unit_list[nx][ny].appendleft(unit_index)

    return len(unit_list[nx][ny]) >= 4


def get_target(unit_index, x, y):

    units = unit_list[x][y]
    target = deque()

    while units and units[0] != unit_index:
        target.append(units.popleft())

    target.append(units.popleft())
    return target


def solve(unit, unit_index):
    x, y, dir = unit.x, unit.y, unit.dir
    nx = x + dx[dir]
    ny = y + dy[dir]

    target = get_target(unit_index, x, y)

    if is_blocked(nx, ny):
        dir = dir + 1 if dir % 2 == 0 else dir - 1
        nx = x + dx[dir]
        ny = y + dy[dir]
        unit_arr[unit_index].set_dir(dir)

        if is_blocked(nx, ny):
            return horses_move(target, x, y, False)

    return horses_move(target, nx, ny, board[nx][ny] == RED)


n, k = map(int, input().split())
board = [list(map(int, input().split())) for _ in range(n)]
unit_list = [[deque() for _ in range(n)] for _ in range(n)]
unit_arr = []

for i in range(k):
    x, y, d = map(int, input().split())
    x -= 1
    y -= 1
    d -= 1

    unit_arr.append(Unit(x, y, d))
    unit_list[x][y].append(i)

turns = 0

while turns <= 1000:
    turns += 1
    for i in range(k):
        if solve(unit_arr[i], i):
            print(turns)
            exit()

print(-1)
